### PR TITLE
refactor: wrote log_assert macro inline in primitives and remove o11y dependency

### PR DESCRIPTION
**Pull request by jaswinder6991**
This is part of the ongoing effort towards reducing dependencies in near-primitives.
See https://github.com/near/nearcore/issues/8888 for context.

The new crate near-log-utils is a light weight crate which has the log_assert and log_assert_fail macros that are used in primitives helping in dropping the o11y dependency from primitives completely.

### CODE REVIEW

**Comment by jaswinder6991** [comment]
This PR is inspired from a similar [PR](#8894).

I have also gone through the Zulip [chat](https://near.zulipchat.com/#narrow/stream/295302-general/topic/publishing.20nearcore.20crates/near/411608655) as mentioned in #8888.

It mentions writing tests for this as well, quoting below:
"To ensure it does not re-occur, writing unit tests and/or comptime assertions is ideal. A heuristic along the lines of "near-primitives has at most 10 transitive dependencies" and such is pretty reasonable, and is also quite straightforward to implement too.
Something like this:
```
cargo tree -p near-primitives --edges=normal | grep -v '*' | wc -l
614
```
"

Being new to the codebase and also to keep the PR simple, I did not look into it.

File: utils/log-assert/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "near-log-utils"
+version.workspace = true
+authors.workspace = true
+edition.workspace = true
+description = "Macros for conditional logging and assertion, offering error handling and diagnostics in both debug and release builds"
+rust-version.workspace = true
+repository.workspace = true
+license.workspace = true
+publish = false
```
Comment (frol) [inline utils/log-assert/Cargo.toml:10]: It has to be publishable so dependent crates can be published

```suggestion
publish = true
```

**Comment by frol** [comment]
@nagisa @Ekleog-NEAR where would be the best place to add the tests for the number of dependencies? (See our previous discussion on Zulip mentioned above)

**Comment by nagisa** [comment]
For one or two uses of the macro in the entire near_primitives I would perhaps opt for inlining the `log_assert` into callsites. The macro is not super complicated, and we avoid all the extra effort that having another publishable micro crate entails.

**Comment by nagisa** [comment]
I mean, a micro crate isn't particularly horrible so far as the maintenace burden is zero. If we could publish this crate once and forget it exists, I'd be plenty fine with approaching it that way.

But this is not the case in our instance. Additional published crates in nearcore do carry a continued maintenance burden. We have plenty of all sorts of checks and lints (in e.g. themis) that this small crate would be subject to and every time these lints become more rigorous, this crate would potentially require additional changes; a new version release every cycle; etc.

> Since it is far from a single place

There are other uses of this macro (and o11y) across the project, correct. However `near-primitives` specifically has exactly a single use of this macro. Here's what that use could look open-coded:

```diff
-log_assert!(
-    epoch_first_header.epoch_id() == header.epoch_id(),
-    "We can only correctly reconstruct headers from this epoch"
-);
+if epoch_first_header.epoch_id() != header.epoch_id() {
+    let msg = "We can only correctly reconstruct headers from this epoch";
+    debug_assert!(false, "{}", msg);
+    tracing::error!(message = msg);
+}
```

Perfectly readable, and not many more new lines, even.

> where would be the best place to add the tests for the number of dependencies?

`themis` is our all-things package-lints dump ground.

**Comment by frol** [comment]
> And as far as I can tell once we do this we can even stop publishing near-o11y again. (And add a comment indicating that it ought to never be published again too.)

This is great!

> Here's what that use could look open-coded:

Yeah, agree, not terrible at all.

@jaswinder6991 Please, do as @nagisa suggested:

* [ ] Inline the macro usage in near-primitives
* [ ] Drop o11y dependency in near-primitives
* [ ] Change `publish` to `false` in the Cargo.toml of o11y and leave a comment there, so future contributors don't suddenly re-enable it
* [ ] Implement the mentioned test inside `themis`

**Comment by jaswinder6991** [comment]
> @jaswinder6991 Actually, I have just noticed that you cloned the implementation instead of extracting it from o11y. I don't suggest to maintain several copies of the macro, and it is used across core and chain crates:
> 
> image
> @nagisa Since it is far from a single place, and I could not think of a better place to put this crate, I think it is fine to have it in a micro-crate for now. I also don't like these function-size crates, so if you have a better solution, please, share.

@frol yes, that was sort of intentional, I did not want to touch anything else other than primitives for now, hence I did not take out the macro from o11y. I can do that it that's the consensus. 

> Perfectly readable, not many more new lines and very little space to introduce tricky bugs.

I can include this change and push it if that is the consensus @nagisa @frol 

> themis is our all-things package-lints dump ground.

Should I look into it for this PR or can this be done later?

**Comment by nagisa** [comment]
> Should I look into it for this PR or can this be done later?

A couple smaller PRs would be nicer to be honest. The one with macro change can be a straightforward slam dunk and is not necessarily tied to changing themis (unless we add a lint specifically about o11y being publishable.)

**Comment by jaswinder6991** [comment]
Thanks @nagisa, I have pushed the changes.

I can submit another PR for the themis tests later. 

Let me know if you have more feedback.

**Comment by frol** [comment]
Well, those two crates will also need to drop the dependency on o11y:

> (i) These private crates break publishable crates. Either make these private crates publishable or avoid using them in the publishable crates:
>  ↳ core/o11y/Cargo.toml (depended on by near-chain-configs and near-dyn-configs)

**Comment by nagisa** [comment]
> Well, those two crates will also need to drop the dependency on o11y:

Aw, that's unfortunate. Their use of o11y is quite fundamental, and unfortunately it looks like these crates are being used by things like jsonrpc in turn. It would perhaps be easier to then figure out a way to make `o11y` not pull the heavy deps that it does. Lets drop the `publish = false` part of the PR for now. Hopefully it is still enough for you and the themis lint would hopefully prevent it becoming a dependency of primitives et al. again.

**Review by nagisa — APPROVED** [review]

**Comment by jaswinder6991** [comment]
> > Well, those two crates will also need to drop the dependency on o11y:
> 
> Aw, that's unfortunate. Their use of o11y is quite fundamental (e.g. near-dyn-config is a companion crate to o11y in the first place), and unfortunately it looks like these crates are being used by things like jsonrpc in turn. It would perhaps be easier to then figure out a way to make `o11y` not pull the heavy deps that it does. Lets drop the `publish = false` part of the PR for now. Hopefully it is still enough for you and the themis lint would hopefully prevent it becoming a dependency of primitives et al. again.

Have made publishable true again for o11y. 
Will create a different PR for the themis test.

**Review by nagisa — APPROVED** [review]

**Comment by jaswinder6991** [comment]
@nagisa I can see that the unused dependencies check failed.
```
unused dependencies:
`near-primitives v0.0.0 (/home/runner/work/nearcore/nearcore/core/primitives)`
└─── dependencies
     └─── "time"
Note: These dependencies might be used by other targets.
      To find dependencies that are not used by any target, enable `--all-targets`.
Note: They might be false-positive.
      For example, `cargo-udeps` cannot detect usage of crates that are only used in doc-tests.
      To ignore some dependencies, write `package.metadata.cargo-udeps.ignore` in Cargo.toml.
error: Recipe `check-cargo-udeps` failed on line 137 with exit code 1
Error: Process completed with exit code 1.
```

Should I remove the time dependency from primitives? It is not directly used in crate. I am not sure if it was there for o11y.
```
time.workspace = true
```

**Comment by nagisa** [comment]
> Should I remove the time dependency from primitives

Lets try removing it first, see what the CI says and if it fails to build, then we'd ignore the dependency. There are false positives and negatives in this check sometimes, and unfortunately the best way to verify I found is to just let CI check it (or you could do `just test` locally too)

**Pull request merged**